module/x/peggy/types/oracle.go

```diff
@@ -0,0 +1,116 @@
+	return sdk.Uint64ToBigEndian(s)
+}
+
+func (n Nonce) String() string {
+	return string(n)
+}
+
+type AttestationCertainty uint8
+
+const (
+	CertaintyUnknown   AttestationCertainty = 0
+	CertaintyRequested AttestationCertainty = 1
+	CertaintyObserved  AttestationCertainty = 2
+)
+
+type AttestationProcessStatus uint8
+
+const (
+	ProcessStatusUnknown   AttestationProcessStatus = 0
+	ProcessStatusInit      AttestationProcessStatus = 1
+	ProcessStatusProcessed AttestationProcessStatus = 2 // prevent double processing
+	ProcessStatusTimeout   AttestationProcessStatus = 3 // end block process will set this
+)
+
+type AttestationProcessResult uint8
+
+const (
+	ProcessResultUnknown AttestationProcessResult = 0
+	ProcessResultSuccess AttestationProcessResult = 1
+	ProcessResultFailure AttestationProcessResult = 2
+)
+
+// Attestation is an aggregate of `claims` that eventually becomes `observed` by all orchestrators
+type Attestation struct {
+	ClaimType           ClaimType
+	Nonce               Nonce // or bytes or int?
+	Certainty           AttestationCertainty
+	Status              AttestationProcessStatus
+	ProcessResult       AttestationProcessResult
+	Tally               AttestationTally
+	SubmitTime          time.Time
+	ConfirmationEndTime time.Time // votes collected <= end time. should be < unbonding period
+	// ExpiryTime time.Time // todo: do we want to keep Attestations forever persisted or can we delete them?
+}
+
+type AttestationTally struct {
+	TotalVotesPower    uint64 // can this overflow?
+	TotalVotesCount    uint64
+	RequiredVotesPower uint64 // todo: revisit if the assumption is true that we can use the values from first claim timestamp
+	RequiredVotesCount uint64 // todo: revisit
+}
+
+func (t *AttestationTally) addVote(power uint64) {
+	t.TotalVotesCount += 1
+	t.TotalVotesPower += power
+}
+
+// ThresholdsReached returns true when votes power > 66% of the validators AND total votes > 50% of the validator count
+func (t AttestationTally) ThresholdsReached() bool {
+	return t.TotalVotesPower > t.RequiredVotesPower &&
+		t.TotalVotesCount > t.RequiredVotesCount
+}
+
+func (a *Attestation) AddVote(now time.Time, power uint64) error {
+	if a.Status != ProcessStatusInit {
+		return sdkerrors.Wrapf(ErrInvalid, "%d", a.Status) // no status to string impl, yet
+	}
+	if now.After(a.ConfirmationEndTime) {
+		return ErrTimeout
+	}
+	a.Tally.addVote(power)
+	if a.Tally.ThresholdsReached() {
+		a.Certainty = CertaintyObserved
+	}
+	return nil
+}
+
+// ID is the unique identifiert used in DB
+func (a *Attestation) ID() []byte {
+	return GetAttestationKey(a.ClaimType, a.Nonce)
+}
+
+// The Fraction type represents a numerator and denominator to enable higher precision thresholds in
+// the election rules. For example:
+// numerator: 1, denominator: 2 => > 50%
+// numerator: 2, denominator: 3 => > 66.666..%
+// numerator: 6273, denominator: 10000 => > 62.73%
+// Valid range of the fraction is 0.5 to 1.
+type Fraction struct {
```

Comment:
Is higher precision really needed? If we have the vote to within 100/2^32 of a percent I think that's more than sufficient.

The extra complexity of this fraction type is IMO a larger risk than a vote failing because 0.000000023% of the voting power opposed it.

And that's if we use u32 and not u256! At that point  the resolution of the vote exceeds the resolution of the native denom on the hub. Since the unit is micro (1*10^-9) rather than wei (1*10^-18)

Reply:
I am not dogmatic with this. The code was copied from the spike. Refactoring this to a `%` based solution is simple.